### PR TITLE
#329 Create CI/CD Pipeline

### LINKED ISSUE #329: Setup Github Actions for CI and pull request.
Setting Github Actions to build the project and run all tests after posting a PR. 
https://docs.github.com/en/actions/language-and-framework-guides/using-nodejs-with-github-actions 
When test failed, there should be a warning on the PR page.

### PR DESCRIPTION
**Pull request by ghost**
**Summary of changes:**

1. Create Build Stage of CI/CD Pipeline.
Notes:
Pushing commits (or doing pull requests) to the server project will trigger server pipeline.
Likewise, pushing commits (or doing pull requests) to the client project will trigger client pipeline.
2. Update package.json of client project with basic `lint` command.
3. Configure Loopback API to use environmental variables for database configuration.
4. Write sample integration test for species controller to utilize environmental variables.

### CODE REVIEW

File: server/.gitignore [hunk 1]
```diff
@@ -63,4 +63,4 @@ typings/
 # Cache used by TypeScript's incremental build
 *.tsbuildinfo
 
-src/datasources/treetracker.datasource.json
+# src/datasources/treetracker.datasource.json
```
Comment (nmcharlton) [inline server/.gitignore:66]: This doesn't look right to me - we shouldn't be committing this file.

Comment (nmcharlton) [reply server/.gitignore:66]: Just spotted the env var bits 😃

File: server/src/datasources/treetracker.datasource.json [hunk 2]
```diff
@@ -0,0 +1,10 @@
+{
```
Comment (dadiorchen) [inline server/src/datasources/treetracker.datasource.json:1]: We shouldn't upload this

Comment (dadiorchen) [reply server/src/datasources/treetracker.datasource.json:1]: And I'm thinking if we need to upload a test datasource file for running test on the Github, maybe we should change the name to another? Maybe there are some risks that we accidentally run the test code locally and do some operation on our real database.

Comment (ghost) [reply server/src/datasources/treetracker.datasource.json:1]: Without the treetracker.datasource.json file in the remote repository, GitHub Actions gives me an error when running `npm run build` for the Loopback project. This is the error I got. 

`##[error]src/datasources/treetracker.datasource.ts(3,25): error TS2307: Cannot find module './treetracker.datasource.json'`

There are several options we can look into when running test code, without doing operation on the actual databases. 

- Using an in-memory database. 
- Connect to a local database (not the dev DB). Afterwards, create the PostgreSQL service container in the CI/CD pipeline.

Comment (nmcharlton) [reply server/src/datasources/treetracker.datasource.json:1]: @dadiorchen
Can we rework the datasource mechanism to use an env var to determine whether or how to read `treetracker.datasource.json`?
For example, we currently choose between that file and `treetrackerTest.datasource.json`, depending on the value of `process.env.NODE_DB`.

We could either:

- Use an env var to specify the location of the file (defaulting to the current location) and add a version of the file to the CI environment, or
- Not read the file at all if `DB_URL` is set (my preference)

Either way, the file should not be tracked in the git repo.

Comment (nmcharlton) [reply server/src/datasources/treetracker.datasource.json:1]: Also, we're importing this file directly in a number of places, which we should avoid.

File: .github/workflows/Admin_CI_Pipeline.yml [hunk 3]
```diff
@@ -0,0 +1,44 @@
+# This workflow will do a clean install of node dependencies, build the source code and run tests across different versions of node
```
[the remark this answers is not in the document]
Comment (ghost) [reply .github/workflows/Admin_CI_Pipeline.yml]: Updated the file to `Admin_Client.yml`.

File: .github/workflows/Admin_Server.yml [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+      working-directory: ${{env.project-directory}}
+    - name: run ESLint
+      run: npm run lint
+      working-directory: ${{env.project-directory}}
+    - name: build server project
+      run: npm run build
+      working-directory: ${{env.project-directory}}
+    - name: run Loopback tests
+      run: npm test
+      working-directory: ${{env.project-directory}}
```
Comment (ghost) [inline .github/workflows/Admin_Server.yml:42]: 1. I removed secrets (for database configuration) in server workflow. 
2. I also removed `continue-on-error`. `continue-on-error` allows the job to continue and pass, even if there are errors that occurred.

File: .github/workflows/Admin_Client.yml [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+# For more information see: https://help.github.com/actions/language-and-framework-guides/using-nodejs-with-github-actions
+
+name: Admin Panel_Client CI/CD Pipeline
+
+on:
+  push:
+    paths:
+    - 'client/**'
+    branches:
+      - master
```
Comment (nmcharlton) [inline .github/workflows/Admin_Client.yml:11]: We should run this workflow on all branches

File: .github/workflows/Admin_Server.yml [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+# For more information see: https://help.github.com/actions/language-and-framework-guides/using-nodejs-with-github-actions
+
+name: Admin Panel_Server CI/CD Pipeline
+
+on:
+  push:
+    paths:
+    - 'server/**'
+    branches:
+      - master
```
Comment (nmcharlton) [inline .github/workflows/Admin_Server.yml:11]: We should run this workflow on all branches

File: .github/workflows/Admin_Server.yml [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+    - name: npm clean install
+      run: npm ci
+      working-directory: ${{env.project-directory}}
+    - name: run ESLint
+      run: npm run lint
+      working-directory: ${{env.project-directory}}
+    - name: build server project
+      run: npm run build
+      working-directory: ${{env.project-directory}}
+    - name: run Loopback tests
```
Comment (nmcharlton) [inline .github/workflows/Admin_Server.yml:40]: These tests currently require a database connection; we can omit these from the workflow for the moment.

File: server/src/__tests__/species.controller.ts [hunk 8]
```diff
@@ -0,0 +1,29 @@
+import { expect } from '@loopback/testlab';
+import { TreetrackerDataSource } from '../datasources';
+import { SpeciesController } from '../controllers';
+import { SpeciesRepository } from '../repositories';
+require('dotenv').config();
+
+/*Purpose: Include env. variables in test in order to 
+map to env. variables in Github Actions.*/
+describe('Species Controller Integration Tests', () => {
+    //Configure the database with env. variables.
```
Comment (nmcharlton) [inline server/src/__tests__/species.controller.ts:10]: While I agree it's useful to be able to configure the DB connection in the environment, the responsibility to read the env vars and set up the datasource object shouldn't sit with the test.

**Comment by nmcharlton** [comment]
Let's get the client and server workflows in and running static and unit tests for now.
We can deal with DB connection issues separately, possibly as part of the integration test issue (#330).

**Comment by nmcharlton** [comment]
I'm going to close this PR and pull the workflow elements across to a new branch to allow me to address the comments and make use of the GitHub Actions UI in my own fork.

**Pull request closed without merging**